# syntax-transform/3.1/0001-prelude.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# This patch adds a mechanism to selectively enable syntax transform in only
# some Ruby scripts.
#
# * When the Ruby functions `eval` or `instance_eval` are called while
#   `mkxp_syntax_transform_enabled_for_next_eval` is true,
#   `mkxp_syntax_transform_enabled_for_next_eval` is immediately set to false
#   and then `mkxp_parser_enable_syntax_transform()` is called on the parser
#   object used by that `eval` or `instance_eval` call to flag the parser as
#   having syntax transform enabled. The value of
#   `mkxp_syntax_transform_enabled_for_next_eval` can be changed from C/C++ to
#   specify which top-level `eval` calls need to have syntax transform enabled.
# * ASTs output by parsers on which `mkxp_parser_enable_syntax_transform()` has
#   been called have a flag set that indicates the AST has syntax transform
#   enabled.
# * Instruction sequences derived from ASTs that are flagged as having syntax
#   transform enabled are themselves flagged as having syntax transform enabled.
# * The syntax transform flag propagates to child instruction sequences of
#   instruction sequences.
# * When the Ruby functions `eval` or `instance_eval` are called from an
#   instruction sequence that has been flagged as having syntax transform
#   enabled, `mkxp_parser_enable_syntax_transform()` is called on the parser
#   object used by that `eval` or `instance_eval` call to flag the parser as
#   having syntax transform enabled. This ensures that the syntax transform flag
#   propagates into `eval` and `instance_eval` calls.
# * The other patches may add code that checks if the syntax transform flag is
#   set on the current parser or instruction sequence, depending on what part of
#   the codebase is being modified, in order to ensure that the syntax transform
#   is only enabled on the Ruby scripts where it should be enabled.

--- a/compile.c
+++ b/compile.c
@@ -44,6 +44,23 @@
 #include "insns.inc"
 #include "insns_info.inc"
 
+/**
+ * Check if mkxp syntax transform is active in the scope that `iseq` is in
+ * and is currently targeting a Ruby version that is less than or equal to
+ * the version given by `maximum_target_ruby_version_major` and
+ * `maximum_target_ruby_version_minor`
+ */
+static bool
+mkxp_is_syntax_transform_active(const rb_iseq_t *iseq, unsigned int maximum_target_ruby_version_major, unsigned int maximum_target_ruby_version_minor)
+{
+    extern unsigned int mkxp_syntax_transform_target_ruby_version_major, mkxp_syntax_transform_target_ruby_version_minor;
+    if (mkxp_syntax_transform_target_ruby_version_major > maximum_target_ruby_version_major || (mkxp_syntax_transform_target_ruby_version_major == maximum_target_ruby_version_major && mkxp_syntax_transform_target_ruby_version_minor > maximum_target_ruby_version_minor)) {
+        return FALSE;
+    }
+
+    return iseq->body->param.flags.mkxp_syntax_transform_flag;
+}
+
 #undef RUBY_UNTYPED_DATA_WARNING
 #define RUBY_UNTYPED_DATA_WARNING 0
 
--- a/iseq.c
+++ b/iseq.c
@@ -613,6 +613,8 @@ prepare_iseq_build(rb_iseq_t *iseq,
     VALUE err_info = Qnil;
     struct rb_iseq_constant_body *const body = iseq->body;
 
+    body->param.flags.mkxp_syntax_transform_flag = option->mkxp_syntax_transform_flag;
+
     if (parent && (type == ISEQ_TYPE_MAIN || type == ISEQ_TYPE_TOP))
 	err_info = Qfalse;
 
@@ -767,6 +769,7 @@ set_compile_option_from_hash(rb_compile_option_t *option, VALUE opt)
     SET_COMPILE_OPTION(option, opt, frozen_string_literal);
     SET_COMPILE_OPTION(option, opt, debug_frozen_string_literal);
     SET_COMPILE_OPTION(option, opt, coverage_enabled);
+    SET_COMPILE_OPTION(option, opt, mkxp_syntax_transform_flag);
     SET_COMPILE_OPTION_NUM(option, opt, debug_level);
 #undef SET_COMPILE_OPTION
 #undef SET_COMPILE_OPTION_NUM
--- a/iseq.h
+++ b/iseq.h
@@ -222,6 +222,7 @@ struct rb_compile_option_struct {
     unsigned int frozen_string_literal: 1;
     unsigned int debug_frozen_string_literal: 1;
     unsigned int coverage_enabled: 1;
+    unsigned int mkxp_syntax_transform_flag: 1;
     int debug_level;
 };
 
--- a/parse.y
+++ b/parse.y
@@ -330,6 +330,8 @@ struct parser_params {
     unsigned int error_p: 1;
     unsigned int cr_seen: 1;
 
+    unsigned int mkxp_syntax_transform_flag: 1;
+
 #ifndef RIPPER
     /* Ruby core only */
 
@@ -359,6 +361,23 @@ struct parser_params {
 #endif
 };
 
+/**
+ * Check if mkxp syntax transform is active in the current parser
+ * and is currently targeting a Ruby version that is less than or equal to
+ * the version given by `maximum_target_ruby_version_major` and
+ * `maximum_target_ruby_version_minor`
+ */
+static bool
+mkxp_is_syntax_transform_active(struct parser_params *p, unsigned int maximum_target_ruby_version_major, unsigned int maximum_target_ruby_version_minor)
+{
+    extern unsigned int mkxp_syntax_transform_target_ruby_version_major, mkxp_syntax_transform_target_ruby_version_minor;
+    if (mkxp_syntax_transform_target_ruby_version_major > maximum_target_ruby_version_major || (mkxp_syntax_transform_target_ruby_version_major == maximum_target_ruby_version_major && mkxp_syntax_transform_target_ruby_version_minor > maximum_target_ruby_version_minor)) {
+        return FALSE;
+    }
+
+    return p->mkxp_syntax_transform_flag;
+}
+
 #define intern_cstr(n,l,en) rb_intern3(n,l,en)
 
 #define STR_NEW(ptr,len) rb_enc_str_new((ptr),(len),p->enc)
@@ -8209,6 +8228,17 @@ parser_set_compile_option_flag(struct parser_params *p, const char *name, const
 		 RBOOL(b));
 }
 
+#ifndef RIPPER
+void
+mkxp_parser_enable_syntax_transform(VALUE vparser)
+{
+    struct parser_params *p;
+    TypedData_Get_Struct(vparser, struct parser_params, &parser_data_type, p);
+    p->mkxp_syntax_transform_flag = TRUE;
+    parser_set_compile_option_flag(p, "mkxp_syntax_transform_flag", "true");
+}
+#endif
+
 static void
 parser_set_shareable_constant_value(struct parser_params *p, const char *name, const char *val)
 {
--- a/vm_core.h
+++ b/vm_core.h
@@ -400,6 +400,8 @@ struct rb_iseq_constant_body {
 	    unsigned int ambiguous_param0 : 1; /* {|a|} */
 	    unsigned int accepts_no_kwarg : 1;
             unsigned int ruby2_keywords: 1;
+
+            unsigned int mkxp_syntax_transform_flag: 1;
 	} flags;
 
 	unsigned int size;
--- a/vm_eval.c
+++ b/vm_eval.c
@@ -1,3 +1,13 @@
+RUBY_FUNC_EXPORTED unsigned int mkxp_syntax_transform_target_ruby_version_major = -1, mkxp_syntax_transform_target_ruby_version_minor = -1;
+
+RUBY_FUNC_EXPORTED
+#if __STDC_VERSION__ >= 202311L
+thread_local
+#else
+_Thread_local
+#endif
+int mkxp_syntax_transform_enabled_for_next_eval = 0;
+
 /**********************************************************************
 
   vm_eval.c -
@@ -1704,6 +1714,19 @@ eval_make_iseq(VALUE src, VALUE fname, int line, const rb_binding_t *bind,
         fname = eval_default_path;
     }
 
+    /* If eval() or instance_eval() is called while mkxp_syntax_transform_enabled_for_next_eval is true, enable syntax transform for this call and immediately set mkxp_syntax_transform_enabled_for_next_eval to false */
+    if (mkxp_syntax_transform_enabled_for_next_eval) {
+        void mkxp_parser_enable_syntax_transform(VALUE);
+        mkxp_syntax_transform_enabled_for_next_eval = 0;
+        mkxp_parser_enable_syntax_transform(parser);
+    }
+
+    /* If code with syntax transform enabled calls eval() or instance_eval(), propagate the syntax transform into the eval() or instance_eval() call */
+    if (parent->body->param.flags.mkxp_syntax_transform_flag) {
+        void mkxp_parser_enable_syntax_transform(VALUE);
+        mkxp_parser_enable_syntax_transform(parser);
+    }
+
     rb_parser_set_context(parser, parent, FALSE);
     ast = rb_parser_compile_string_path(parser, fname, src, line);
     if (ast->body.root) {
--- a/vm_insnhelper.c
+++ b/vm_insnhelper.c
@@ -31,6 +31,23 @@
 #include "insns_info.inc"
 #endif
 
+/**
+ * Check if mkxp syntax transform is active in the scope that `iseq` is in
+ * and is currently targeting a Ruby version that is less than or equal to
+ * the version given by `maximum_target_ruby_version_major` and
+ * `maximum_target_ruby_version_minor`
+ */
+static bool
+mkxp_is_syntax_transform_active(const rb_iseq_t *iseq, unsigned int maximum_target_ruby_version_major, unsigned int maximum_target_ruby_version_minor)
+{
+    extern unsigned int mkxp_syntax_transform_target_ruby_version_major, mkxp_syntax_transform_target_ruby_version_minor;
+    if (mkxp_syntax_transform_target_ruby_version_major > maximum_target_ruby_version_major || (mkxp_syntax_transform_target_ruby_version_major == maximum_target_ruby_version_major && mkxp_syntax_transform_target_ruby_version_minor > maximum_target_ruby_version_minor)) {
+        return FALSE;
+    }
+
+    return iseq->body->param.flags.mkxp_syntax_transform_flag;
+}
+
 extern rb_method_definition_t *rb_method_definition_create(rb_method_type_t type, ID mid);
 extern void rb_method_definition_set(const rb_method_entry_t *me, rb_method_definition_t *def, void *opts);
 extern int rb_method_definition_eq(const rb_method_definition_t *d1, const rb_method_definition_t *d2);
